ibm/resource_ibm_kms_key_rings.go

```diff
@@ -0,0 +1,248 @@
+)
+
+func resourceIBMKmskeyRings() *schema.Resource {
+	return &schema.Resource{
+		Create:   resourceIBMKmsKeyRingCreate,
+		Delete:   resourceIBMKmsKeyRingDelete,
+		Read:     resourceIBMKmsKeyRingRead,
+		Importer: &schema.ResourceImporter{},
+
+		Schema: map[string]*schema.Schema{
+			"instance_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Key ID",
+				ForceNew:    true,
+			},
+			"key_ring_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Key protect or hpcs key alias name",
+			},
+			"endpoint_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateAllowedStringValue([]string{"public", "private"}),
+				Description:  "public or private",
+				ForceNew:     true,
+				Default:      "public",
+			},
+		},
+	}
+}
+
+func resourceIBMKmsKeyRingCreate(d *schema.ResourceData, meta interface{}) error {
+	kpAPI, err := meta.(ClientSession).keyManagementAPI()
+	if err != nil {
+		return err
+	}
+	rContollerClient, err := meta.(ClientSession).ResourceControllerAPIV2()
+	if err != nil {
+		return err
+	}
+
+	instanceID := d.Get("instance_id").(string)
+	endpointType := d.Get("endpoint_type").(string)
+	keyRingID := d.Get("key_ring_id").(string)
+	log.Println("****instance id:****", instanceID)
+
+	rContollerAPI := rContollerClient.ResourceServiceInstanceV2()
+
+	instanceData, err := rContollerAPI.GetInstance(instanceID)
+	if err != nil {
+		return err
+	}
+	instanceCRN := instanceData.Crn.String()
+	crnData := strings.Split(instanceCRN, ":")
+
+	var hpcsEndpointURL string
+
+	if crnData[4] == "hs-crypto" {
+		hpcsEndpointAPI, err := meta.(ClientSession).HpcsEndpointAPI()
+		if err != nil {
+			return err
+		}
+
+		resp, err := hpcsEndpointAPI.Endpoint().GetAPIEndpoint(instanceID)
+		if err != nil {
+			return err
+		}
+
+		if endpointType == "public" {
+			hpcsEndpointURL = "https://" + resp.Kms.Public + "/api/v2/keys"
+		} else {
+			hpcsEndpointURL = "https://" + resp.Kms.Private + "/api/v2/keys"
+		}
+
+		u, err := url.Parse(hpcsEndpointURL)
+		if err != nil {
+			return fmt.Errorf("Error Parsing hpcs EndpointURL")
+		}
+		kpAPI.URL = u
+	} else if crnData[4] == "kms" {
+		if endpointType == "private" {
+			if !strings.HasPrefix(kpAPI.Config.BaseURL, "private") {
+				kpAPI.Config.BaseURL = "private." + kpAPI.Config.BaseURL
+			}
+		}
+	} else {
+		return fmt.Errorf("Invalid or unsupported service Instance")
+	}
+	kpAPI.Config.InstanceID = instanceID
+
+	err = kpAPI.CreateKeyRing(context.Background(), keyRingID)
+	if err != nil {
+		return fmt.Errorf(
+			"Error while creating key ring : %s", err)
+	}
+	var keyRing string
+	keyRings, err2 := kpAPI.GetKeyRings(context.Background())
+	if err2 != nil {
+		return fmt.Errorf(
+			"Error while fetching key ring : %s", err2)
+	}
+	for _, v := range keyRings.KeyRings {
+		log.Print(v)
+		if v.ID == d.Get("key_ring_id") {
```

Comment:
Instead of d.Get("key_ring_id").. Use already stored keyRingID

Reply:
Need Discussion